pkg/cluster/fixssh.go

```diff
@@ -124,52 +161,117 @@ func (m *manager) removeBackendPoolsFromNIC(ctx context.Context, resourceGroup,
-	ilbID := fmt.Sprintf("%s/backendAddressPools/%s", *lb.ID, ilbBackendPool)
+	sshBackendPoolID := fmt.Sprintf("%s/backendAddressPools/ssh-%d", *lb.ID, i)
+	ilbBackendPoolID := fmt.Sprintf("%s/backendAddressPools/%s", *lb.ID, ilbBackendPool)
 
 	updateSSHPool := true
 	updateILBPool := true
 
-	for _, p := range *(*nic.IPConfigurations)[0].LoadBalancerBackendAddressPools {
-		if strings.EqualFold(*p.ID, sshID) {
-			updateSSHPool = false
-		}
-		if strings.EqualFold(*p.ID, ilbID) {
-			updateILBPool = false
+	ipc := (*nic.InterfacePropertiesFormat.IPConfigurations)[0]
+	if ipc.LoadBalancerBackendAddressPools == nil {
+		emptyBackendAddressPool := make([]mgmtnetwork.BackendAddressPool, 0)
+		(*nic.IPConfigurations)[0].LoadBalancerBackendAddressPools = &emptyBackendAddressPool
+	} else {
+		for _, p := range *(*nic.IPConfigurations)[0].LoadBalancerBackendAddressPools {
+			if strings.EqualFold(*p.ID, sshBackendPoolID) {
+				updateSSHPool = false
+			}
+			if strings.EqualFold(*p.ID, ilbBackendPoolID) {
+				updateILBPool = false
+			}
 		}
 	}
 
 	if updateSSHPool {
-		m.log.Printf("Adding NIC %s to Internal Load Balancer SSH Backend Address Pool %s", nicName, sshID)
+		m.log.Printf("Adding NIC %s to Internal Load Balancer SSH Backend Address Pool %s", nicName, sshBackendPoolID)
 		*(*nic.IPConfigurations)[0].LoadBalancerBackendAddressPools = append(*(*nic.IPConfigurations)[0].LoadBalancerBackendAddressPools, mgmtnetwork.BackendAddressPool{
-			ID: &sshID,
+			ID: &sshBackendPoolID,
 		})
 	}
 
 	if updateILBPool {
-		m.log.Printf("Adding NIC %s to Internal Load Balancer API Address Pool %s", nicName, ilbID)
+		m.log.Printf("Adding NIC %s to Internal Load Balancer API Address Pool %s", nicName, ilbBackendPoolID)
 		*(*nic.IPConfigurations)[0].LoadBalancerBackendAddressPools = append(*(*nic.IPConfigurations)[0].LoadBalancerBackendAddressPools, mgmtnetwork.BackendAddressPool{
-			ID: &ilbID,
+			ID: &ilbBackendPoolID,
 		})
 	}
 
 	return updateSSHPool || updateILBPool
 }
 
+func (m *manager) updateV1ELBAddressPool(ctx context.Context, nic *mgmtnetwork.Interface, nicName string, lb *mgmtnetwork.LoadBalancer, i int, infraID string) bool {
+	if nic.InterfacePropertiesFormat.IPConfigurations == nil || len(*nic.InterfacePropertiesFormat.IPConfigurations) == 0 {
+		m.log.Warnf("unable to update NIC as there are no IP configurations for %s", nicName)
+		return false
+	}
+	var updateV1ELBPool bool
+
+	elbBackendPoolID := fmt.Sprintf("%s/backendAddressPools/%s", *lb.ID, infraID)
+	lbpool := *(*nic.IPConfigurations)[0].LoadBalancerBackendAddressPools
+	var index int
+	for i, p := range lbpool {
+		if strings.EqualFold(*p.ID, elbBackendPoolID) {
+			updateV1ELBPool = true
+			index = i
+		}
+	}
+
+	if updateV1ELBPool {
+		m.log.Printf("Removing NIC %s from Public Load Balancer API Address Pool %s", nicName, elbBackendPoolID)
+		lbpool = append(lbpool[:index], lbpool[index+1:]...)
+		(*nic.IPConfigurations)[0].LoadBalancerBackendAddressPools = &lbpool
```

Comment:
lets refactor this to avoid side effects. we can move
```
m.log.Printf("Updating Network Interface %s", nicName)
		err = m.interfaces.CreateOrUpdateAndWait(ctx, resourceGroup, nicName, *nic)
		if err != nil {
			return err
		}
```
into this condition and then we won't be relying on side effects to call update.

Reply:
Refactored the code as per your suggestion